RecoLocalTracker/SiPixelRecHits/src/PixelCPETemplateReco.cc

```diff
@@ -543,3 +543,14 @@ LocalError PixelCPETemplateReco::localError(DetParam const& theDetParam, Cluster
 
   return LocalError(xerr * xerr, 0, yerr * yerr);
 }
+
+void PixelCPETemplateReco::fillPSetDescription(edm::ParameterSetDescription& desc) {
+  desc.add<int>("barrelTemplateID", 0);
+  desc.add<int>("forwardTemplateID", 0);
+  desc.add<int>("directoryWithTemplates", 0);
+  desc.add<int>("speed", -2);
+  desc.add<bool>("UseClusterSplitter", false);
+
+  // needed because it's in the HLT configuration
+  desc.addOptional<bool>("DoCosmics", false);
```

Comment:
If this is really not needed, I wouldn't propagate it in the code indefinitely.
HLT can be customized for removing it from its config, and then once this new code will be parsed in confDB the customization will be removed.
Please remove here and add the customization in HLTrigger/Configuration/python/customizeHLTforCMSSW.py

Reply:
@perrotta,
somewhat inspired by what was done in #19098  I tried naively:
```diff
diff --git a/HLTrigger/Configuration/python/customizeHLTforCMSSW.py b/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
index 1c58edf2ae6..e78325866dc 100644
--- a/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
+++ b/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
@@ -46,6 +46,10 @@ def customiseFor2017DtUnpacking(process):
 
     return process
 
+def customiseFor27694(process):
+    for producer in producers_by_type(process, "PixelCPETemplateRecoESProducer"):
+        if hasattr(producer, "DoCosmics"): del producer.DoCosmics
+    return process
 
 # CMSSW version specific customizations
 def customizeHLTforCMSSW(process, menuType="GRun"):
@@ -53,4 +57,6 @@ def customizeHLTforCMSSW(process, menuType="GRun"):
     # add call to action function in proper order: newest last!
     # process = customiseFor12718(process)
 
+    process = customiseFor27694(process)
+
     return process
diff --git a/RecoLocalTracker/SiPixelRecHits/src/PixelCPETemplateReco.cc b/RecoLocalTracker/SiPixelRecHits/src/PixelCPETemplateReco.cc
index c13fdc90e8c..dccc01ebfbe 100644
--- a/RecoLocalTracker/SiPixelRecHits/src/PixelCPETemplateReco.cc
+++ b/RecoLocalTracker/SiPixelRecHits/src/PixelCPETemplateReco.cc
@@ -552,5 +552,5 @@ void PixelCPETemplateReco::fillPSetDescription(edm::ParameterSetDescription& des
   desc.add<bool>("UseClusterSplitter", false);
 
   // needed because it's in the HLT configuration
-  desc.addOptional<bool>("DoCosmics", false);
+  //desc.addOptional<bool>("DoCosmics", false);
 }
```

but this fails addOnTests...
Please advice if there is any obvious mistake on my side.

Reply:
OK, I figured out one should use `esproducers_by_type` instead.
Your comment has been addressed (together with some other clean-up) in https://github.com/cms-sw/cmssw/pull/27694/commits/12ce9d6eb35b21aebee9925edf88805797ef7f02.